Map multi valued ldap field to column in user table

mbeh-ub opened the pull request:
Hi,

if using an LDAP for authentication config.ini allows to map fields from LDAP-Schema to fields in the user-Table. At the moment only single-valued fields are supported. This little code snippet checks if the ldap-field is multi-valued and serializes the data if necessary.
Are you interested in such an extension?

demiankatz commented:
A couple of concerns here:

1.) Based on a read of the code, it appears that this is going to change/break the existing behavior. The existing logic seems to assume that $value is going to be an array containing a single element. The new logic will serialize all arrays, whether or not they contain multiple values. This means that something like a mapping for a name is going to result in a serialized value rather than the first element of the array. (Apologies if I'm mistaken and am misreading something).

2.) I imagine that we might want some more flexibility here. Serialization is useful if we want to be able to rebuild data structures for processing, but many of the values being extracted in here are intended for end-user display purposes (names, addresses, etc.). It seems to me in that case, we would be more likely to want either the existing behavior (take the first value) or something more display-oriented (apply a delimiter to a list). Perhaps this is something that should be somehow configurable on a field-by-field basis. What do you think?

mbeh-ub commented:
Good points. I fixed issue 1) and added an option "separator" to the LDAP-section in config.ini.
If a separator string is given multiple-values would be mapped to the given user attribute. If no separator is given the old behavior should be kept.

demiankatz commented:
Looks like good progress! A couple more questions/comments:

1.) There seem to be some spelling inconsistencies that will probably result in bugs (seperator vs. separator).

2.) I see that you are unsetting a special 'count' property. Is there any danger of other special properties causing weird behaviors? Do we need to filter on numeric keys to be completely safe?

mbeh-ub commented:
fixed point 1)
I don't think ther are other non-numeric properties except the 'count'-property. 
The array ist returned by the php function http://php.net/manual/en/function.ldap-get-entries.php. 
And should have only 'count' as non-numeric property.
But to be on the save side I updated the code and iterate explicity over the numeric values (from 0 to count -1).

demiankatz commented:
Looks good! I made some minor adjustments to comments and merged this to master. Thanks!

A couple of thoughts for possible future enhancements, which we can live without but which might be worth adding if you need them for your own use cases:

1.) Right now, the list of user table fields is hard-coded. If you wanted to add a custom field to your user table, you would have to extend the LDAP auth class. It might be helpful to have an "extra user fields" setting in the config to extend the built-in values.

2.) Right now, when separator is enabled, ALL fields will get delimited. It might be useful to have a "delimited_fields" setting to optionally choose to delimit only some fields, while taking first values for others.

In any case, thanks again for this enhancement, and please feel no obligation to do more work unless you actually need to. Just thinking of possibilities, not trying to create work for the sake of work. :-)
